### PR TITLE
Convert Bundler lockfile checksum validation to manual opt-in

### PR DESCRIPTION
**Pull request by martinemde**
## What was the end-user or developer problem that led to this PR?

I've not yet been able to finish the full-featured opt-in for checksums.

## What is your fix for the problem, implemented in this PR?

This makes bundler checksums a hidden opt-in feature. There are no commands which add the CHECKSUMS section to the lockfile and the feature will only be enabled if that section is found.

The actual behavior change is as follows:
* When checksums come from either the API or from a gem, we add them to the checksum store unless the settings disabled checksum validation.
* When checksums come from the lockfile, we always store them in the checksum store regardless of the disabled checksums setting.
* The presence of a checksum section, even empty, in the lockfile indicates that we should write a checksum section to the lockfile when we generate.
* When the disable checksums setting is set, we still read and write checksums to the lockfile, but since all other checksums won't be recorded, we will never compare them or raise an error.
* When the checksums section is not in the lockfile, the api and gem checksums will continue to be stored and compared, perpetuating the existing behavior of checksum verification with the server, but not writing any of them to the lockfile.
* Adding an empty CHECKSUMS section to the lockfile will enable the feature manually. We will follow up with more user-friendly commands to add and populate this new section.

## Make sure the following tasks are checked

- [X] Describe the problem / feature
- [X] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [X] Write code to solve the problem
- [X] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

File: bundler/lib/bundler/lockfile_parser.rb [hunk 1]
```diff
@@ -226,10 +229,7 @@ def parse_checksum(line)
       platform = $4
 
       version = Gem::Version.new(version)
-      platform = platform ? Gem::Platform.new(platform) : Gem::Platform::RUBY
```
Comment (segiddins) [inline bundler/lib/bundler/lockfile_parser.rb:229]: why is this line removed?

File: bundler/lib/bundler/lockfile_parser.rb [hunk 2]
```diff
@@ -111,6 +111,9 @@ def initialize(lockfile)
         elsif line == DEPENDENCIES
           @parse_method = :parse_dependency
         elsif line == CHECKSUMS
+          # This is a temporary solution to make this feature disabled by default
+          # for all gemfiles that don't already explicitly include the feature.
+          Bundler::Checksum.enable!
```
Comment (segiddins) [inline bundler/lib/bundler/lockfile_parser.rb:116]: this feels super brittle to me, given bundler could have read a different lockfile and then have stored the ivar.

I would feel a lot better if this got hoisted into the definition

File: bundler/lib/bundler/checksum.rb [hunk 3]
```diff
@@ -9,6 +9,23 @@ class Checksum
...
     class << self
+      # This is used by the lockfile parser to enable checksum validation
+      # when checksums are found in the lockfile.
+      # We don't enable if the user has explicitly disabled checksum validation.
+      # TODO: This can be removed when checksums are enabled by default.
+      # The enable_checksum_validation setting may also be removed once this
+      # is opt-out instead of opt-in
+      def enable!
+        return if Bundler.settings[:disable_checksum_validation]
+        @enabled = true
```
Comment (segiddins) [inline bundler/lib/bundler/checksum.rb:20]: will this ivar ever get cleared?

File: bundler/lib/bundler/lockfile_parser.rb (hunk 1 above)
Comment (martinemde) [reply bundler/lib/bundler/lockfile_parser.rb:229]: oh, it's duplicating the behavior of Gem::NameTuple.new so it's unnecessary. Maybe it can happen in another PR, but it's something I noticed when going through the code.

File: bundler/lib/bundler/lockfile_parser.rb (hunk 2 above)
Comment (martinemde) [reply bundler/lib/bundler/lockfile_parser.rb:116]: ok, I'll shift it there. Would you still use the same ivar approach? I toyed with adding a Bundler.setting for "enabled", but it makes it more complex. Do you prefer that?

File: bundler/lib/bundler/lockfile_parser.rb (hunk 1 above)
Comment (martinemde) [reply bundler/lib/bundler/lockfile_parser.rb:229]: I should say, not duplicating exactly, it's making a Gem::Platform, then to_s on the platform. The only reason this would matter is if there's an inconsistent platform name saved into the checksums section of the lockfile that gets resolved by passing through `Gem::Platform.new#to_s`. I'll revert it just based on this slim case.

File: bundler/lib/bundler/checksum.rb [hunk 4]
```diff
@@ -9,6 +9,29 @@ class Checksum
...
+      # TODO: This can be removed when checksums are enabled by default.
+      # The enable_checksum_validation setting may also be removed once this
+      # is opt-out instead of opt-in
+      def enabled?
+        !Bundler.settings[:disable_checksum_validation] &&
+          Bundler.settings[:enable_checksum_validation]
+      end
+
+      def from_gem_package(gem_package, algo = DEFAULT_ALGORITHM)
+        return nil unless enabled?
```
Comment (segiddins) [inline bundler/lib/bundler/checksum.rb:24]: this is a behavior change - we should still be validating the checksums from the server on install by default (unless `:disable_checksum_validation` is set)

Comment (martinemde) [reply bundler/lib/bundler/checksum.rb:13]: well, I suppose it wasn't lost until we began disabling this feature. Now we need to add a special case for it.

File: bundler/spec/cache/gems_spec.rb [hunk 5]
```diff
@@ -289,22 +289,49 @@
       expect(cached_gem("rack-1.0.0")).to exist
     end
 
-    it "raises an error when the gem file is altered and produces a different checksum" do
+    it "raises an error when the gem is altered and produces a different checksum" do
+      pending "This test depends on lockfile checksums being enabled by default"
```
Comment (martinemde) [inline bundler/spec/cache/gems_spec.rb:293]: I don't understand why this one won't trigger the error. I enable checksums in the lockfile.

File: bundler/spec/install/gemfile/platform_spec.rb [hunk 6]
```diff
@@ -224,21 +233,16 @@
...
+      #{checksums}
       BUNDLED WITH
          #{Bundler::VERSION}
     L
 
     bundle "lock --add-platform ruby"
 
+    good_checksums = checksums.dup
+    good_checksums.no_checksum("pry", "0.11.3")
+
```
Comment (martinemde) [inline bundler/spec/install/gemfile/platform_spec.rb:245]:
```suggestion
```

This isn't needed any more.

File: bundler/spec/install/gemfile/sources_spec.rb [hunk 7]
```diff
@@ -379,14 +396,11 @@
                   rack
 
             PLATFORMS
-              #{lockfile_platforms}
+              #{lockfile_platforms("ruby")}
```
Comment (martinemde) [inline bundler/spec/install/gemfile/sources_spec.rb:399]: Locally I had to add ruby to get this passing, now it's failing on windows.

File: bundler/spec/install/gemfile/sources_spec.rb [hunk 8]
```diff
@@ -422,14 +436,11 @@
...
```
Comment (martinemde) [inline bundler/spec/install/gemfile/sources_spec.rb:439]: Same problem here.

File: bundler/spec/install/gemfile/sources_spec.rb (hunk 7 above)
Comment (deivid-rodriguez) [reply bundler/spec/install/gemfile/sources_spec.rb:399]: In most situations `lockfile_platforms` should include `ruby` by default since #5700. What's the error if you revert the change? I can't reproduce it locally.

File: bundler/spec/cache/gems_spec.rb (hunk 5 above)
Comment (deivid-rodriguez) [reply bundler/spec/cache/gems_spec.rb:293]: I think it may be because we currently don't produce the error either if the checksum section is there, but there was no locked checksum for the gem that was altered (like in the lockfile you added to the spec).

If I modify the spec like this, it passes:

```diff
diff --git a/bundler/spec/cache/gems_spec.rb b/bundler/spec/cache/gems_spec.rb
index 2f5da4e7e..bb07bfc08 100644
--- a/bundler/spec/cache/gems_spec.rb
+++ b/bundler/spec/cache/gems_spec.rb
@@ -289,15 +289,32 @@
       expect(cached_gem("rack-1.0.0")).to exist
     end
 
-    it "raises an error when the gem file is altered and produces a different checksum" do
+    it "raises an error when the gem is altered and produces a different checksum" do
       cached_gem("rack-1.0.0").rmtree
       build_gem "rack", "1.0.0", :path => bundled_app("vendor/cache")
+
+      expected_checksums = checksums_section do |c|
+        c.checksum gem_repo1, "actionpack", "2.3.2"
+        c.checksum gem_repo1, "activesupport", "2.3.2"
+        c.checksum gem_repo1, "rack", "1.0.0"
+      end
+
       simulate_new_machine
 
+      lockfile <<~L
+        GEM
+          remote: #{file_uri_for(gem_repo2)}/
+          specs:
+            rack (1.0.0)
+
+        CHECKSUMS
+          #{expected_checksums}
+      L
+
       bundle :install, :raise_on_error => false
-      expect(exitstatus).to eq(37)
       expect(err).to include("Bundler found mismatched checksums.")
       expect(err).to include("1. remove the gem at #{cached_gem("rack-1.0.0")}")
+      expect(exitstatus).to eq(37)
 
       expect(cached_gem("rack-1.0.0")).to exist
       cached_gem("rack-1.0.0").rmtree
```

Conversely, if I add the lockfile that you added here to the spec in current master branch, it no longer raises the error.

**Comment by martinemde** [comment]
@deivid-rodriguez Updated with your suggestions. I don't know what happened with the ruby platform there, but if I remove it, it works now. This should be updated and ready to go, if you approve.

File: bundler/spec/cache/gems_spec.rb (hunk 5 above)
Comment (martinemde) [reply bundler/spec/cache/gems_spec.rb]: Thanks for figuring this out. It makes sense that it needs the checksum stored in the lockfile to trigger this. It probably was inserting the checksum automatically when they were enabled by default so I didn't realize it was necessary for the feature to work.

**Comment by martinemde** [comment]
I updated the description of this PR to include the new behavior. I'm happy with this approach because it means the feature is present and ready to use, even can be enabled to work exactly as merged initially, but it is disabled. We can then add new commands that add the section and populate it and those can go out in point releases because they don't change any behavior without explicit interaction from the user. Any 2.5 series bundler will be able to handle lockfile checksum and the future commands will just be a bonus.

File: bundler/spec/bundler/definition_spec.rb [hunk 9]
```diff
@@ -56,6 +56,11 @@
         s.add_dependency "rack", "1.0"
       end
 
+      checksums = checksums_section_when_existing do |c|
+        c.no_checksum "foo", "1.0"
+        c.checksum gem_repo1, "rack", "1.0.0"
+      end
+
```
Comment (deivid-rodriguez) [inline bundler/spec/bundler/definition_spec.rb:63]: This should go _after_ the `bundle install` command below. Otherwise it will generate an empty section once we generate checksums by default and will make the spec fail.

File: bundler/spec/lock/lockfile_spec.rb [hunk 10]
```diff
@@ -78,9 +79,6 @@
       DEPENDENCIES
         rack
 
-      CHECKSUMS
-        #{checksum_for_repo_gem(gem_repo2, "rack", "1.0.0")}
-
```
Comment (deivid-rodriguez) [inline bundler/spec/lock/lockfile_spec.rb:83]: Should this include the checksums section so that specs keep passing once generated by default?

File: bundler/spec/lock/lockfile_spec.rb [hunk 11]
```diff
@@ -214,9 +210,6 @@
       DEPENDENCIES
         rack (> 0)
 
-      CHECKSUMS
-        #{checksum_for_repo_gem(gem_repo2, "rack", "1.0.0")}
-
```
Comment (deivid-rodriguez) [inline bundler/spec/lock/lockfile_spec.rb:219]: Should this include the checksums section so that specs keep passing once generated by default?

**Review by deivid-rodriguez — APPROVED** [review]
Looks great, this is fantastic work and leaves master in a much better place towards confidently releasing.

I tried one experiment, which is applying this diff:

```diff
diff --git a/bundler/lib/bundler/definition.rb b/bundler/lib/bundler/definition.rb
index 3493f0732..6018d1b82 100644
--- a/bundler/lib/bundler/definition.rb
+++ b/bundler/lib/bundler/definition.rb
@@ -93,7 +93,7 @@ def initialize(lockfile, dependencies, sources, unlock, ruby_version = nil, opti
         @locked_bundler_version = @locked_gems.bundler_version
         @locked_ruby_version = @locked_gems.ruby_version
         @originally_locked_specs = SpecSet.new(@locked_gems.specs)
-        @locked_checksums = @locked_gems.checksums
+        @locked_checksums = true
 
         if unlock != true
           @locked_deps    = @locked_gems.dependencies
```

I _think_ the way you designed this PR is so that this kind of change would trigger just a few selected spec failures, but I'm seeing quite a few specs fail.

I started commenting on some of them, but then I thought maybe you prefer to merge and then work on those? If we fix most of them, would it make sense to potentially run a CI job under an ENV setting that enables the setting and make sure we don't regress any thing related to checksum generation. Maybe overkill?

**Comment by martinemde** [comment]
That definitely makes sense as a part of CI, imho. It was my intention to leave the tests in a state where we could enable or disable the feature with a switch. I had planned to revisit this after I got the test passing with opt-in. The thing I most wamted to avoid was ruining all the work to add checksums to all the lockfiles in tests 🙀

I will revisit after merge so that we're less blocked than with the feature enabled by default. I suspect failures are not actually the feature being broken, but tests that assert incorrectly or setup incorrectly.

File: bundler/spec/support/checksums.rb [hunk 12]
```diff
@@ -3,53 +3,82 @@
...
-    def checksum_section(&block)
-      ChecksumsBuilder.new(&block).to_lock
+    def checksums_section(enabled = true, &block)
+      ChecksumsBuilder.new(enabled, &block)
     end
 
-    def checksum_for_repo_gem(*args)
-      checksum_section do |c|
-        c.repo_gem(*args)
+    def checksums_section_when_existing(&block)
```
[the remark this answers is not in the document]
Comment (martinemde) [reply bundler/spec/support/checksums.rb:57]: That's the intention, though as we discussed I think in practice there's some differences. I'll revisit in the next PR.

File: bundler/spec/commands/lock_spec.rb [hunk 13]
```diff
@@ -65,42 +62,61 @@
...
-    expect(read_lockfile).to eq(remove_checksums_from_lockfile(@lockfile))
+    expect(read_lockfile).to eq(@lockfile)
+  end
+
+  it "prints a lockfile without fetching new checksums if the existing lockfile had no checksums" do
+    lockfile remove_checksums_from_lockfile(@lockfile)
+
+    bundle "lock --print"
+
+    expect(out).to eq(remove_checksums_from_lockfile(@lockfile).chomp)
```
[the remark this answers is not in the document]
Comment (martinemde) [reply bundler/spec/commands/lock_spec.rb:95]: Oh probably, they used to express different behaviors like auto-adding checksums, but now they've flattened to be the same.

File: bundler/spec/install/gemfile/platform_spec.rb [hunk 14]
```diff
@@ -264,16 +265,7 @@
       DEPENDENCIES
         empyrean (= 0.1.0)
         pry
-
-      CHECKSUMS
-        #{checksum_for_repo_gem gem_repo4, "coderay", "1.1.2"}
-        #{checksum_for_repo_gem gem_repo4, "empyrean", "0.1.0"}
-        #{checksum_for_repo_gem gem_repo4, "ffi", "1.9.23", "java"}
-        #{checksum_for_repo_gem gem_repo4, "method_source", "0.9.0"}
-        pry (0.11.3)
```
[the remark this answers is not in the document]
Comment (martinemde) [reply bundler/spec/install/gemfile/platform_spec.rb:273]: Oops, this is a missed conversion.

**Comment by deivid-rodriguez** [comment]
Yep, the failures I looked at all seemed like a problem with the specs, not with the feature. Good to merge away and iterate 👍.

File: bundler/spec/commands/lock_spec.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (martinemde) [reply bundler/spec/commands/lock_spec.rb:95]: I fixed this one and re-pushed.
